### PR TITLE
Warn if a # is specified in a break point URL(Issue #4277)

### PR DESCRIPTION
**Pull request by varshav21**

### CODE REVIEW

File: src/org/zaproxy/zap/extension/brk/impl/http/BreakAddEditDialog.java [hunk 1]
```diff
@@ -146,6 +146,9 @@ public String validateFields() {
 			}
 		}
 		return null;
+		if(FIELD_STRING.contains("#")==true){
```
Comment (thc202) [inline src/org/zaproxy/zap/extension/brk/impl/http/BreakAddEditDialog.java]: - It should do the contains on the value of the field (e.g. `getStringValue(FIELD_STRING)`).
 - The comparison to `true` could be removed.
 - Note that this check needs to be done before the above return statement.

File: src/org/zaproxy/zap/extension/brk/impl/http/BreakAddEditDialog.java [hunk 2]
```diff
@@ -146,6 +146,9 @@ public String validateFields() {
...
+			return Constant.messages.getString("brk.brkpoint.edit.title");
```
Comment (thc202) [inline src/org/zaproxy/zap/extension/brk/impl/http/BreakAddEditDialog.java]: It needs to be a new/custom message that indicates the actual problem.

**Comment by kingthorin** [comment]
@varshav21 are you able to address the feedback?

**Comment by varshav21** [comment]
Thank you so much, yes I am working on it.Do I add a new custom key in messages.properties file for the return message?

**Comment by kingthorin** [comment]
Yes. Something like `brk.brkpoint.warn.urlfragment=Pattern shouldn't include URL fragment (#)` or something along those lines.

**Comment by varshav21** [comment]
Why did the Travis CI build fail when I pushed commits to this branch but was successful when I pushed the same commits to the master branch and started a separate pull request?

**Comment by thc202** [comment]
This one has different file changes (note that the class has leading plus chars in some lines), removing them should pass the build.

**Comment by thc202** [comment]
BTW, the new validation should also check if the `FIELD_LOCATION` is `url`(?) and `Match` is `contains`?

**Comment by varshav21** [comment]
Should I take the key for url and contains from the messages.properties file and check if they are equal to strings FIELD_LOCATION and FIELD_MATCH respectively in the if condition?

**Comment by thc202** [comment]
For that we can use the methods `strToLoc` and `strToMatch`, then compare against the `Location.url` and `Match.contains`.

File: src/org/zaproxy/zap/extension/brk/impl/http/BreakAddEditDialog.java [hunk 3]
```diff
@@ -145,7 +145,10 @@ public String validateFields() {
 				return Constant.messages.getString("brk.brkpoint.error.regex"); 
 			}
 		}
-		return null;
+		if( (this.getStringValue(FIELD_STRING)).contains("#") && (Location.url.equals(this.strToLoc(this.getStringValue(FIELD_LOCATION)))) && (Match.contains.equals(this.strToMatch(this.getStringValue(FIELD_MATCH)))) ){
```
Comment (thc202) [inline src/org/zaproxy/zap/extension/brk/impl/http/BreakAddEditDialog.java]: Some of the parentheses could be removed (e.g. the ones surrounding `this.getStringValue(FIELD_STRING)` and following operands).

File: src/org/zaproxy/zap/extension/brk/impl/http/BreakAddEditDialog.java [hunk 4]
```diff
@@ -145,7 +145,10 @@ public String validateFields() {
...
+			return Constant.messages.getString("brk.brkpoint.warn.urlfragment");
+		}
+		return null;
```
Comment (thc202) [inline src/org/zaproxy/zap/extension/brk/impl/http/BreakAddEditDialog.java]: The tabs added at the end of the line should be removed.

**Comment by thc202** [comment]
Thank you!

**Comment by thc202** [comment]
@varshav21 how would you like to be credited?
https://github.com/zaproxy/zap-core-help/wiki/HelpCredits

**Comment by varshav21** [comment]
Thank you very much for your guidance. Varsha Venkatesh is fine.

**Pull request merged**